stream/stream.go

```diff
@@ -0,0 +1,135 @@
+package stream
+
+import (
+	"encoding/gob"
+	"errors"
+	"net/http"
+
+	"github.com/google/zoekt"
+	"github.com/google/zoekt/query"
+	"github.com/google/zoekt/rpc"
+)
+
+const DefaultSSEPath = "/stream"
+
+func Server(searcher zoekt.Searcher) http.Handler {
+	rpc.RegisterGob()
+	return &streamHandler{Searcher: searcher}
+}
+
+type searchArgs struct {
+	Q    query.Q
+	Opts *zoekt.SearchOptions
+}
+
+type searchReply struct {
+	Event  string
+	Result *zoekt.SearchResult
+}
+
+type streamHandler struct {
+	Searcher zoekt.Searcher
+}
+
+func (h *streamHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+
+	// Decode payload.
+	args := new(searchArgs)
+	err := gob.NewDecoder(r.Body).Decode(args)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	eventWriter, err := newEventStreamWriter(w)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	// Always send a done event in the end.
+	defer func() {
+		err = eventWriter.event("done", nil)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+		}
+	}()
+
+	// Kick-off search (in batch-mode for now).
+	searchResults, err := h.Searcher.Search(ctx, args.Q, args.Opts)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	// We simulate streaming by sending searchResults in chunks over the wire. Later
+	// we want Searcher.Search to take a channel, buffer here and send chunks over
+	// the network.
+	var chunk *zoekt.SearchResult
+	// We send stats first. We don't send RepoURLs or LineFragments because we don't
+	// use them in Sourcegraph.
+	chunk = &zoekt.SearchResult{
+		Stats: searchResults.Stats,
+	}
+	// Send event.
+	err = eventWriter.event("matches", chunk)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
```

Comment:
we should send an error event. By this point we may have already sent other headers/data down so setting the http status is too late.

Reply:
Errors cannot be gob encoded. We can send a string instead and recreate the error on the other side.